---

title: Iterative_masking


keywords: fastai
sidebar: home_sidebar

summary: "Supporting repository for: "Generative power of a protein language model trained on multiple sequence alignments" (preprint: https://doi.org/10.1101/2022.04.14.488405). We use MSA Transformer (https://doi.org/10.1101/2021.02.12.430858) to generate synthetic protein sequences by masking iteratively the same MSA."
description: "Supporting repository for: "Generative power of a protein language model trained on multiple sequence alignments" (preprint: https://doi.org/10.1101/2022.04.14.488405). We use MSA Transformer (https://doi.org/10.1101/2021.02.12.430858) to generate synthetic protein sequences by masking iteratively the same MSA."
nb_path: "index.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: index.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Getting-started">Getting started<a class="anchor-link" href="#Getting-started"> </a></h2><p>Clone this repository on your local machine by running:</p>
<div class="highlight"><pre><span></span>git clone git@github.com:Bitbol-Lab/Iterative_masking.git
</pre></div>
<p>and move inside the root folder.
One can the use directly the functions from the cloned repository or install it with an editable install using:</p>
<div class="highlight"><pre><span></span>pip install -e .
</pre></div>
<p>We recommend creating and activating a dedicated <code>conda</code> or <code>virtualenv</code> Python virtual environment.</p>
<h2 id="Requirements">Requirements<a class="anchor-link" href="#Requirements"> </a></h2><p>In order to use the functions, the following python packages are required:</p>
<ul>
<li>numpy</li>
<li>scipy</li>
<li>numba</li>
<li>fastcore</li>
<li>biopython</li>
<li>esm==0.4.0</li>
<li>pytorch</li>
</ul>
<p>It is also required to use a GPU (with cuda).</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="How-to-use">How to use<a class="anchor-link" href="#How-to-use"> </a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><a href="/Iterative_masking/core.html#IM_MSA_Transformer"><code>IM_MSA_Transformer</code></a>: Class with different functions used to generate new MSAs with the iterative masking procedure</p>
<p><a href="/Iterative_masking/core.html#gen_MSAs"><code>gen_MSAs</code></a>: example function (with parser) that can be used to generate and save new sequences directly from the terminal.</p>

</div>
</div>
</div>
</div>
 

